pennylane/spin/lattice.py

```diff
@@ -0,0 +1,196 @@
+from scipy.spatial import cKDTree
+
+# pylint: disable=too-many-arguments, too-many-instance-attributes
+# pylint: disable=use-a-generator, too-few-public-methods
+
+
+class Lattice:
+    r"""Constructs a Lattice object.
+
+    Args:
+       L: Number of unit cells in each direction.
+       unit_cell: Primitive vectors for the lattice.
+       basis: Initial positions of spins.
+       boundary_condition: defines boundary conditions, boolean or series of bools with dimensions same as L.
+       neighbour_order: Range of neighbours a spin interacts with.
+       distance_tol: Distance below which spatial points are considered equal for the purpose of identifying nearest neighbours.
+
+    Returns:
+       Lattice object
+    """
+
+    def __init__(
+        self,
+        L,
+        unit_cell,
+        basis=None,
+        boundary_condition=False,
+        neighbour_order=1,
+        distance_tol=1e-5,
+    ):
+
+        self.L = np.asarray(L)
+        self.n_dim = len(L)
+        self.unit_cell = np.asarray(unit_cell)
+        if basis is None:
+            basis = np.zeros(self.unit_cell.shape[0])[None, :]
+
+        self.basis = np.asarray(basis)
+        self.n_sl = len(self.basis)
+        self.n_sites = np.prod(L) * self.n_sl
+
+        if isinstance(boundary_condition, bool):
+            boundary_condition = [boundary_condition for _ in range(self.n_dim)]
+
+        self.boundary_condition = boundary_condition
+        self._test_input_accuracy()
+
+        self.lattice_points, lattice_map = self._generate_grid(neighbour_order)
+
+        cutoff = neighbour_order * np.linalg.norm(self.unit_cell, axis=1).max() + distance_tol
+        edges = self._identify_neighbours(cutoff)
+        self._generate_true_edges(edges, lattice_map, neighbour_order)
+
+    def _test_input_accuracy(self):
+        r"""Tests the accuracy of the input provided"""
+
+        for l in self.L:
+            if (not isinstance(l, np.int64)) or l <= 0:
+                raise TypeError("Argument `L` must be a list of positive integers")
+
+        if self.unit_cell.ndim != 2:
+            raise ValueError("'unit_cell' must have ndim==2, as array of primitive vectors.")
+
+        if self.basis.ndim != 2:
+            raise ValueError("'basis' must have ndim==2, as array of initial coordinates.")
+
+        if self.unit_cell.shape[0] != self.unit_cell.shape[1]:
+            raise ValueError("The number of primitive vectors must match their length")
+
+        if not all(isinstance(b, bool) for b in self.boundary_condition):
+            raise ValueError(
+                "Argument 'boundary_condition' must be a bool or a list of bools of same dimensions as the unit_cell"
+            )
+
+        if len(self.boundary_condition) != self.n_dim:
+            raise ValueError(
+                "Argument 'boundary_condition' must be a bool or a list of bools of same dimensions as the unit_cell"
+            )
+
+    def _identify_neighbours(self, cutoff):
```

Comment:
Can you please add the `bigO` complexity for this function's algorithm?

Reply:
Do you want this added to the docstring or just want me to mention it here?

Reply:
It could be left as a comment for fellow developers.